dali/python/nvidia/dali/_utils/external_source_impl.py

```diff
@@ -255,3 +271,177 @@ def get_callback_from_source(source, cycle):
+            self.iteration += 1
+            return batch_to_numpy(result, _tf_batch_error_msg)
+
+    return CallableBatchIterator, dtype, shape
+
+
+def get_sample_iterable_from_callback(source_desc, batch_size):
+    """Transform sample callback accepting one argument into an Iterable
+    """
+    first = source_desc.source(types.SampleInfo(0, 0, 0))
+    dtype, shape = _inspect_data(first, False)
+
+    class CallableSampleIterator:
+        first_value = first
+        def __init__(self):
+            self.idx_in_epoch = 0
+            self.idx_in_batch = 0
+            self.iteration = 0
+            self.source = source_desc.source
+
+        def __iter__(self):
+            self.idx_in_epoch = 0
+            self.idx_in_batch = 0
+            self.iteration = 0
+            return self
+
+        def __next__(self):
+            if self.idx_in_epoch == 0 and CallableSampleIterator.first_value is not None:
+                result = CallableSampleIterator.first_value
+                CallableSampleIterator.first_value = None
+            else:
+                idx = types.SampleInfo(self.idx_in_epoch, self.idx_in_batch, self.iteration)
+                result = self.source(idx)
+            self.idx_in_epoch += 1
+            self.idx_in_batch += 1
+            if self.idx_in_batch == batch_size:
+                self.idx_in_batch = 0
+                self.iteration += 1
+            return sample_to_numpy(result, _tf_sample_error_msg)
+
+    return CallableSampleIterator, dtype, shape
+
+
+def get_iterable_from_callback(source_desc, is_batched):
+    """Transform callback that doesn't accept arguments into iterable
+    """
+    print("get_iterable_from_callback")
+    first = source_desc.source()
+    dtype, shape = _inspect_data(first, is_batched)
+
+    class CallableIterator:
+        first_value = first
+        def __init__(self):
+            self.source = source_desc.source
+
+        def __iter__(self):
+            return self
+
+        def __next__(self):
+            if CallableIterator.first_value is not None:
+                result = CallableIterator.first_value
+                CallableIterator.first_value = None
+            else:
+                result = self.source()
+            if is_batched:
+                return batch_to_numpy(result, _tf_batch_error_msg)
+            else:
+                return sample_to_numpy(result, _tf_sample_error_msg)
+
+    return CallableIterator, dtype, shape
+
+
+def get_iterable_from_iterable_or_generator(source_desc, is_batched):
+    """Wrap iterable or generator function into another iterable while peeking the first element
+
+    If the source is generator function it must be called first.
+    """
+    if source_desc.kind == SourceKind.GENERATOR_FUNC:
+        first_iter = iter(source_desc.source())
+    else:
+        first_iter = iter(source_desc.source)
+    first =  next(first_iter)
+    dtype, shape = _inspect_data(first, is_batched)
+
+    class PeekFirstGenerator:
```

Comment:
Does this name mean "This class is a generator that peeks first" or "This class peeks the first generator"?

Reply:
Class to wrap generator (in another generator) that used the first peeked value. Maybe not the best name, but it's pretty internal.